ReactNativeClient/lib/models/Tag.js

```diff
@@ -174,6 +256,110 @@ class Tag extends BaseItem {
 		}
 	}
 
+	static tagPath(tags, tagId) {
+		return nestedPath(tags, tagId);
+	}
+
+	static displayTitle(item) {
+		const title = super.displayTitle(item);
+		const separator = '/';
+		const i = title.lastIndexOf(separator);
+		if (i !== -1) return title.slice(i + separator.length);
+		return title;
+	}
+
+	static async changeDescendantPrefix(parentItem, parentPrefix) {
+		const childrenTagIds = await Tag.childrenTagIds(parentItem.id);
+		for (let i = 0; i < childrenTagIds.length; i++) {
+			const childItem = await Tag.load(childrenTagIds[i]);
+			// The new title of the child is new prefix + display name
+			const childName = `${parentPrefix}/${await Tag.displayTitle(childItem)}`;
+			childItem.title = childName;
+			// Change the child title
+			await Tag.save(childItem, { fields: ['title'] });
+			// Change descendant titles
+			await Tag.changeDescendantPrefix(childItem, childName);
+		}
+	}
+
+	static async rename(tag, newTitle) {
+		const oldParentId = tag.parent_id;
+
+		tag.title = newTitle;
+		tag = await Tag.saveNested(tag, { fields: ['title', 'parent_id'], userSideValidation: true });
+		await Tag.changeDescendantPrefix(tag, newTitle);
+
+		if (oldParentId !== tag.parent_id) {
+			// If the parent tag has changed, and the ancestor doesn't
+			// have notes attached, then remove it
+			const oldParentWithCount = await Tag.loadWithCount(oldParentId);
+			if (!oldParentWithCount) {
+				await Tag.delete(oldParentId, { deleteChildren: false, deleteNotelessParents: true });
+			}
+		}
+		return tag;
+	}
+
+	static async moveTag(tagId, parentTagId) {
+		if (tagId === parentTagId
+				|| (await Tag.descendantTagIds(tagId)).includes(parentTagId)) {
+			throw new Error(_('Cannot move tag to this location.'));
```

Comment:
There shouldn't be special methods to change tag properties because they can be modified via the API too for example. It should be possible to call `Tag.save({ title: "abc" })` or `Tag.save({ parent_id: "123" })` and to be confident the app is going to do the right thing. So if there's extra logic that need to happen in this case, it should be moved to Tag.save (preferably under `underSideValidation`.

Also please append a `_` to private methods. You can perhaps use this as private helper methods, that you'd call from Tag.save.

Reply:
The purpose of this method is slightly different now after the latest change - it is used to rename the full_title of a tag, i.e. `a/b/c`->`b/c/d`, it handles the creation of new parent tags (if necessary) and deletes old parents if they're no longer linked to any notes. I've changed the name to `renameNested` to reflect this. 
You can still change the short `title` and `parent_id` using the `Tag.save` method.